src/strategies/offer_maker/market_making/Kandel/Kandel.sol

```diff
@@ -0,0 +1,265 @@
+    Bid,
+    Ask
+  }
+
+  ///@notice maps index to offer id on Mangrove
+  mapping(OrderType => mapping(uint => uint)) offerIdOfIndex;
+
+  constructor(IMangrove mgv, IERC20 base, IERC20 quote, uint gasreq, uint16 nslots) Direct(mgv, NO_ROUTER, gasreq) {
+    NSLOTS = nslots;
+    BASE = base;
+    QUOTE = quote;
+  }
+
+  ///@notice sets the base/quote distribution that Kandel should follow using continuous slices
+  ///@param from start index (included). Must be less than `to`.
+  ///@param to end index (excluded). Must be less than `NSLOT`.
+  ///@param slice slice[i][0/1] is the distribution of base/quote at index i.
+  /// `slice[i][0/1] = 0` is interpreted as keep old distribution
+  function setDistribution(uint from, uint to, uint[2][] calldata slice) external onlyAdmin {
+    for (uint i = from; i < to; i++) {
+      if (slice[i][0] > 0) {
+        baseOfIndex[i] = slice[i][0];
+      }
+      if (slice[i][1] > 0) {
+        quoteOfIndex[i] = slice[i][1];
+      }
+    }
+  }
+
+  ///@notice sets the base/quote distribution that Kandel should follow at specified indexes
+  ///@param indexes the indexes of the distribution one wishes to set
+  ///@param newDistrib `newDistrib[i][0/1]` is the new base/quote distribution at index `indexes[i]`
+  function setDistribution(uint[] calldata indexes, uint[2][] calldata newDistrib) external onlyAdmin {
+    uint cpt = 0;
+    for (uint i = 0; i < indexes.length; i++) {
+      if (newDistrib[cpt][0] > 0) {
+        baseOfIndex[indexes[i]] = newDistrib[cpt][0];
+      }
+      if (newDistrib[cpt][1] > 1) {
+        quoteOfIndex[indexes[i]] = newDistrib[cpt][1];
+      }
+      cpt++;
+    }
+  }
+
+  ///@notice how much price and volume distribution Kandel should give at given index
+  ///@param ba whether Kandel is asking or bidding at this index
+  ///@param index the distribution index
+  function _givesOfIndex(OrderType ba, uint index) internal view returns (uint) {
+    return ba == OrderType.Ask ? quoteOfIndex[index] : baseOfIndex[index];
+  }
+
+  ///@notice how much price and volume distribution Kandel should want at given index
+  ///@param ba whether Kandel is asking or bidding at this index
+  ///@param index the distribution index
+  function _wantsOfIndex(OrderType ba, uint index) internal view returns (uint) {
+    return ba == OrderType.Ask ? baseOfIndex[index] : quoteOfIndex[index];
+  }
+
+  ///@notice turns an order type into an (outbound, inbound) pair identifying an offer list
+  ///@param ba whether one wishes to access the offer lists where asks or bids are posted
+  function _tokenPairOfOrderType(OrderType ba) internal view returns (IERC20, IERC20) {
+    return ba == OrderType.Bid ? (QUOTE, BASE) : (BASE, QUOTE);
+  }
+
+  ///@notice returns the Kandel order type of the offer list whose outbound token is given in argument
+  ///@param outbound_tkn the outbound token of the offer list
+  function _orderTypeOfOutbound(IERC20 outbound_tkn) internal view returns (OrderType) {
```

Comment:
Are you intending to override these methods in specializations?
From our natspec of MangroveOffer:
```
/// `_f() internal`: descendant of this contract should provide a public wrapper for this function, with necessary guards.
```

Normally we do not prefix with _ do we?

Reply:
Yes, I think common practice is to use `_` prefix to denote an internal function. Maybe we should update natspec?